image-processing.js

```diff
@@ -0,0 +1,59 @@
+'use strict'
+const R = require('ramda')
+const stream = require('stream')
+const sharp = require('sharp')
+
+const renameThumb = (name, options) => {
+  const dotIndex = name.lastIndexOf('.')
+  const dotExtension = name.slice(dotIndex)
+  const filename = name.slice(0, dotIndex)
+  const label = options.label || `@${options.width}x${options.height}`
+
+  return `${filename}${label}${dotExtension}`
+}
+
+module.exports = (client, inputArgs, cache) => {
+  const name = inputArgs.name
+  const data = inputArgs.data
+  const options = inputArgs.options
+  let originalData, fullsize
+
+  if ((inputArgs.data instanceof stream.Stream)) {
+    originalData = new stream.PassThrough()
+    data.pipe(originalData)
+    fullsize = originalData.pipe(sharp().resize(options.maxSize, options.maxSize).max().withoutEnlargement().on('info', info => options.meta = info))
+  } else {
+    originalData = data
+    fullsize = sharp(originalData).resize(options.maxSize, options.maxSize).max().withoutEnlargement().on('info', info => options.meta = info)
+  }
+
+  const originalUpload = client.upload(name, fullsize, options).then(cache.put(name, data))
```

Comment:
@JonForest not sure if you are aware but `client.upload` can take the raw data to upload or a stream. In this case a stream. Sharp is also a writable stream. So essentially whats happening is that originalData gets piped into sharp and then inside `client.upload`, the sharp stream is then piped into aws. (I think)